src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs

```diff
@@ -379,6 +379,39 @@ protected override bool ReplacementChangesSemanticsForNodeLanguageSpecific(Synta
                     }
                 }
             }
+            else if (currentOriginalNode.Kind() == SyntaxKind.CaseSwitchLabel)
+            {
+                var originalCaseSwitchLabel = (CaseSwitchLabelSyntax)currentOriginalNode;
+                var newCaseSwitchLabel = (CaseSwitchLabelSyntax)currentReplacedNode;
+
+                // If case label is changing, then need to check if the semantics will change for the switch expression.  
+                // e.g. if switch expression is "switch(x)" where "object x = 1f", then "case 1:" and "case (float) 1:" are different.
+                var originalCaseType = this.OriginalSemanticModel.GetTypeInfo(previousOriginalNode, this.CancellationToken).Type;
+                var newCaseType = this.SpeculativeSemanticModel.GetTypeInfo(previousReplacedNode, this.CancellationToken).Type;
+
+                if (originalCaseType == newCaseType)
+                    return false;
+
+                var oldSwitchStatement = originalCaseSwitchLabel?.Parent?.Parent as SwitchStatementSyntax;
+                var newSwitchStatement = newCaseSwitchLabel?.Parent?.Parent as SwitchStatementSyntax;
+                if (oldSwitchStatement == null || newSwitchStatement == null)
+                {
+                    return false;
+                }
+
+                var originalConversion = this.OriginalSemanticModel.ClassifyConversion(oldSwitchStatement.Expression, originalCaseType);
+                var newConversion = this.SpeculativeSemanticModel.ClassifyConversion(newSwitchStatement.Expression, newCaseType);
+                
+                // if conversion only exists for either original or new, then semantics changed.
+                if (originalConversion.Exists != newConversion.Exists)
+                {
+                    return true;
+                }
```

Comment:
I was thinking of returning `false` if `originalConversion.Exists` and `newConversion.Exists` are both false, but then I believe the resulting type should have been the same in order for the semantic to have remained the same. So I don't think I should return `false` straight away.